src/librados/IoCtxImpl.cc

```diff
@@ -747,6 +747,65 @@ int librados::IoCtxImpl::operate_read(const object_t& oid,
   return r;
 }
 
+int librados::IoCtxImpl::aio_operate_repair_read(const object_t& oid,
+				      ::ObjectOperation *o,
+				      AioCompletionImpl *c,
+				      bufferlist *pbl,
+				      int flags, int32_t osdid, epoch_t e, int op_flags)
+{
+  if (!o->size())
+    return 0;
+
+  Context *onack = new C_aio_Ack(c);
+
+  c->is_read = true;
+  c->io = this;
+
+  version_t ver;
+
+  int op = o->ops[0].op.op;
+  ldout(client->cct, 10) << ceph_osd_op_name(op) << " oid=" << oid << " nspace=" << oloc.nspace << dendl;
+  // Prepend assert_interval op
+  OSDOp tmp;
+  tmp.op.op = CEPH_OSD_OP_ASSERT_INTERVAL;
+  tmp.op.assert_interval.epoch = e;
+  o->ops.insert(o->ops.begin(), tmp);
+  int size = o->ops.size();
+  o->out_bl.resize(size);
+  o->out_handler.resize(size);
+  o->out_rval.resize(size);
+  o->out_bl[1] = pbl;
+  o->ops[1].op.flags = op_flags;
+  Objecter::Op *objecter_op = objecter->prepare_read_op(oid, oloc,
+	                                      *o, snap_seq, NULL,
+	                                      flags | CEPH_OSD_FLAG_REPAIR_READS | CEPH_OSD_FLAG_IGNORE_OVERLAY,
+	                                      onack, &ver);
+  objecter_op->target.osd = osdid;
```

Comment:
at https://github.com/ceph/ceph/blob/master/src/osd/OSD.cc#L8651, reply the client with ENXIO. and at that moment the MOSDOp is only partially decoded, so the MOSDOp::ops is still empty. 

if a non-primary osd is specified, the osd will return a MOSDOpReply with empty ops. on the client side, none of the complete call backs will be called at https://github.com/ceph/ceph/blob/master/src/osdc/Objecter.cc#L3270. this makes the client wait in vain.

@dzafman maybe we can support the ec pool by delegating this call to aio_operate_read() ? or just return an error if the client wants to read a shad by specifying a src osd? or we need to give a revamp of current read path of ec backend to support the (osd,shard). but i wonder what the user would want to do with a erasure shard without the erasure decoder? the data shard (m) might still be recognisable, but what if it's a parity shard (k)?